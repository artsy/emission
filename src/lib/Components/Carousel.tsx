import { Box, color, Flex } from "@artsy/palette"
import OpaqueImageView from "lib/Components/OpaqueImageView"
import React from "react"
import { LayoutChangeEvent, ScrollView, TouchableWithoutFeedback } from "react-native"
import styled from "styled-components/native"

// Hardcoded height from Show carousel:
// https://github.com/artsy/eigen/blob/1526f4dad04742be19ed2fb219f94d0512734326/Artsy/View_Controllers/Fair/ARShowViewController.m#L90
const ITEM_HEIGHT = 250

const ImageView = styled(OpaqueImageView)`
  height: ${ITEM_HEIGHT}px;
`

const PageList = styled(ScrollView)`
  height: ${ITEM_HEIGHT}px;
`

const PageIndicator = styled.View<{ isActive: boolean; isLast: boolean }>`
  width: 6px;
  height: 6px;
  border-radius: 3px;
  background-color: ${({ isActive }) => (isActive ? color("black100") : color("black10"))};
  margin-right: ${({ isLast }) => (isLast ? 0 : 12)}px;
`

interface Props {
  sources: Array<{
    imageURL: string
    aspectRatio: number
  }>
}

interface State {
  pageWidth: number
  activePageIdx: number
  animationTargetIdx: number
}

export class Carousel extends React.Component<Props, State> {
  scrollView: ScrollView
  state = {
    activePageIdx: 0,
    animationTargetIdx: null,
    pageWidth: null,
  }

  handleScroll = ev => {
    const {
      nativeEvent: {
        contentOffset: { x },
      },
    } = ev
    const { pageWidth, animationTargetIdx } = this.state
    const currentPageIdx = Math.round(x <= 0 ? 0 : x / pageWidth)
    if (animationTargetIdx !== null) {
      // Clear indicator "lock" if we've reached our destination, otherwise
      // ignore events triggered by intermediate animated pages.
      if (animationTargetIdx === currentPageIdx) {
        this.setState({ animationTargetIdx: null })
      }
    } else {
      this.setState({
        activePageIdx: currentPageIdx,
      })
    }
  }

  handleLayout = (ev: LayoutChangeEvent) => {
    const {
      nativeEvent: {
        layout: { width },
      },
    } = ev
    if (width && !this.state.pageWidth) {
      this.setState({ pageWidth: width })
    }
  }

  handleIndicatorPress = idx => {
    const { pageWidth } = this.state
    if (this.scrollView) {
      this.setState(
        {
          animationTargetIdx: idx,
          activePageIdx: idx,
        },
        () => {
          this.scrollView.scrollTo({
            x: idx * pageWidth,
            y: 0,
            animated: true,
          })
        }
      )
    }
  }

  keyForSource = ({ imageURL }) => imageURL

  renderItems = () => {
    const { pageWidth } = this.state
    const { sources } = this.props

    if (pageWidth) {
      return sources.map(source => (
        <Flex key={this.keyForSource(source)} width={pageWidth} justifyContent="center" alignItems="center">
          <ImageView {...source} />
        </Flex>
      ))
    }

    // If we're awaiting first layout, render an item with height to prevent vertical reflow jank.
    return <Box height={ITEM_HEIGHT} />
  }

  render() {
    const { sources } = this.props
    const { activePageIdx } = this.state
    return (
      <>
        <PageList
          innerRef={ref => {
            this.scrollView = ref
          }}
          onLayout={this.handleLayout}
          pagingEnabled
          horizontal
          onScroll={this.handleScroll}
          scrollEventThrottle={160}
          showsHorizontalScrollIndicator={false}
        >
          {this.renderItems()}
        </PageList>
        <Flex flexDirection="row" mt={1} mb={1} justifyContent="center" alignItems="center">
          {sources.map((source, idx) => (
            <TouchableWithoutFeedback key={this.keyForSource(source)} onPress={() => this.handleIndicatorPress(idx)}>
              <PageIndicator isActive={idx === activePageIdx} isLast={idx === sources.length - 1} />
            </TouchableWithoutFeedback>
          ))}
        </Flex>
      </>
    )
  }
}
