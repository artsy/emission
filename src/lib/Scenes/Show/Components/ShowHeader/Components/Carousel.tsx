import { Box, Flex, space } from "@artsy/palette"
import OpaqueImageView from "lib/Components/OpaqueImageView"
import React from "react"
import { Dimensions, ScrollView, ScrollViewProperties } from "react-native"
import styled from "styled-components/native"

interface Props extends ScrollViewProperties {
  sources: Array<{
    imageURL: string
    aspectRatio: number
  }>
}

export class Carousel extends React.Component<Props> {
  scrollView: ScrollView

  keyForSource = ({ imageURL }) => imageURL

  renderItems = () => {
    const { sources } = this.props

    return sources.map((source, i) => {
      const lastImage = i === sources.length - 1
      return (
        <Flex key={this.keyForSource(source)} mr={lastImage ? 2 : 1} pr={lastImage ? 2 : 0} alignItems="flex-start">
          <ImageView {...source} isFirst={i === 0} />
        </Flex>
      )
    })
  }

  render() {
    return (
      <Box my={2}>
        <PageList
          {...this.props}
          ref={ref => {
            this.scrollView = ref
          }}
          horizontal
          scrollEventThrottle={160}
          showsHorizontalScrollIndicator={false}
        >
          {this.renderItems()}
        </PageList>
      </Box>
    )
  }
}

const ITEM_HEIGHT = 350

const { width: windowWidth } = Dimensions.get("window")

const ImageView = styled(OpaqueImageView)<{ isFirst: boolean; aspectRatio: number }>`
  height: ${ITEM_HEIGHT};
  margin-top: ${p => (p.isFirst ? 150 : 0)};
  ${p =>
    p.isFirst &&
    `
      height: 200;
      width: ${windowWidth - space(2) - 80};
    `};
`

const PageList = styled(ScrollView)`
  height: ${ITEM_HEIGHT}px;
  padding-left: ${space(2)};
`
