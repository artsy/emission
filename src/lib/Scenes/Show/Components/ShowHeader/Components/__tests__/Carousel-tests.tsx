import React from "react"
import * as renderer from "react-test-renderer"

import { Carousel } from "../Carousel"

describe("Carousel", () => {
  it("renders properly", () => {
    const carousel = renderer.create(<Carousel sources={images} />)
    expect(carousel).toMatchSnapshot()
  })
})

const images = [
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/GZAyAwIBNitUUogXsAoaBw/larger.jpg", aspectRatio: 1.5 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/GkBR99xMC__SWyi4j5kJdw/larger.jpg", aspectRatio: 1.5 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/zDbgyKt6x3yGoUHFkO--Aw/larger.jpg", aspectRatio: 1.5 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/Eejy0FSVyqJ8NlSe8j8j6Q/larger.jpg", aspectRatio: 1.5 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/aLHd1-5iDe3eTcSb04QwtA/larger.jpg", aspectRatio: 0.67 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/66Lg4kZ4tjMD5nGfPOFUyA/larger.jpg", aspectRatio: 1.5 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/WcGcGFg5y4pnqZAacbDu3w/larger.jpg", aspectRatio: 1.5 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/6Oq63DESPbMW4W4mM3b01g/larger.jpg", aspectRatio: 1.5 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/cyb8k0si34ih-Qf_X5m7JA/larger.jpg", aspectRatio: 1.5 },
  { imageURL: "https://d32dm0rphc51dk.cloudfront.net/PGB4v5DaCME2JjlVJclZ8g/larger.jpg", aspectRatio: 1.5 },
]
