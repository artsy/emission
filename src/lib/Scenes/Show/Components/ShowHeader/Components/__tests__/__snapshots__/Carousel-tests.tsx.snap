// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders properly 1`] = `
<View
  className="css-1j9oeic"
  style={
    Array [
      Object {},
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    scrollEventThrottle={160}
    showsHorizontalScrollIndicator={false}
    sources={
      Array [
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/GZAyAwIBNitUUogXsAoaBw/larger.jpg",
        },
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/GkBR99xMC__SWyi4j5kJdw/larger.jpg",
        },
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/zDbgyKt6x3yGoUHFkO--Aw/larger.jpg",
        },
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/Eejy0FSVyqJ8NlSe8j8j6Q/larger.jpg",
        },
        Object {
          "aspectRatio": 0.67,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/aLHd1-5iDe3eTcSb04QwtA/larger.jpg",
        },
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/66Lg4kZ4tjMD5nGfPOFUyA/larger.jpg",
        },
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/WcGcGFg5y4pnqZAacbDu3w/larger.jpg",
        },
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/6Oq63DESPbMW4W4mM3b01g/larger.jpg",
        },
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/cyb8k0si34ih-Qf_X5m7JA/larger.jpg",
        },
        Object {
          "aspectRatio": 1.5,
          "imageURL": "https://d32dm0rphc51dk.cloudfront.net/PGB4v5DaCME2JjlVJclZ8g/larger.jpg",
        },
      ]
    }
    style={
      Array [
        Object {
          "height": 350,
          "paddingLeft": 20,
        },
      ]
    }
  >
    <View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/GZAyAwIBNitUUogXsAoaBw/larger.jpg"
          isFirst={true}
          style={
            Array [
              Object {
                "height": 200,
                "marginTop": 150,
                "width": 650,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/GkBR99xMC__SWyi4j5kJdw/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/zDbgyKt6x3yGoUHFkO--Aw/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/Eejy0FSVyqJ8NlSe8j8j6Q/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={0.67}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/aLHd1-5iDe3eTcSb04QwtA/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/66Lg4kZ4tjMD5nGfPOFUyA/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/WcGcGFg5y4pnqZAacbDu3w/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/6Oq63DESPbMW4W4mM3b01g/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1fcgr5v"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/cyb8k0si34ih-Qf_X5m7JA/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
      <View
        className="css-1ttsi3g"
        style={
          Array [
            Object {
              "display": "flex",
            },
          ]
        }
      >
        <AROpaqueImageView
          aspectRatio={1.5}
          imageURL="https://d32dm0rphc51dk.cloudfront.net/PGB4v5DaCME2JjlVJclZ8g/larger.jpg"
          isFirst={false}
          style={
            Array [
              Object {
                "height": 350,
                "marginTop": 0,
              },
            ]
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
